### PR TITLE
FEAT-#7146: Use BaseQueryCompiler, BasePandasDataset, DataFrame or Series type hints at a high level

### LINKED ISSUE #7146: Use `BaseQueryCompiler`, `BasePandasDataset`, `DataFrame` or `Series` type hints at a high level
To make navigation easier.

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

To facilitate navigation, allows to automatically find places where functions and attributes are defined.

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #7146 
- [ ] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/distributed/dataframe/pandas/partitions.py [hunk 1]
```diff
@@ -90,7 +90,7 @@ def unwrap_partitions(
             f"Only API Layer objects may be passed in here, got {type(api_layer_object)} instead."
         )
 
-    modin_frame = api_layer_object._query_compiler._modin_frame
+    modin_frame = api_layer_object._query_compiler._modin_frame  # type: ignore[attr-defined]
```
Comment (YarShev) [inline modin/distributed/dataframe/pandas/partitions.py:93]: Why is `# type: ignore[attr-defined]` needed?

File: modin/pandas/accessor.py [hunk 2]
```diff
@@ -208,7 +215,9 @@ class ModinAPI:
         Object to operate on.
     """
 
-    def __init__(self, data):
+    _data: Union[DataFrame, Series]
```
Comment (YarShev) [inline modin/pandas/accessor.py:218]: Why did this become a class attribute?

File: modin/pandas/accessor.py [hunk 3]
```diff
@@ -45,20 +51,21 @@ class BaseSparseAccessor(ClassLogger):
         Object to operate on.
     """
 
+    _parent: Union[DataFrame, Series]
```
Comment (YarShev) [inline modin/pandas/accessor.py:54]: Why did this become a class attribute?

File: modin/pandas/base.py [hunk 4]
```diff
@@ -174,6 +177,7 @@ class BasePandasDataset(ClassLogger):
     # Pandas class that we pretend to be; usually it has the same name as our class
     # but lives in "pandas" namespace.
     _pandas_class = pandas.core.generic.NDFrame
+    _query_compiler: BaseQueryCompiler
```
Comment (YarShev) [inline modin/pandas/base.py:180]: Why do we need this as a class attribute?

File: modin/pandas/groupby.py [hunk 5]
```diff
@@ -85,10 +91,12 @@
 class DataFrameGroupBy(ClassLogger):
     _pandas_class = pandas.core.groupby.DataFrameGroupBy
     _return_tuple_when_iterating = False
+    _df: Union[DataFrame, Series]
+    _query_compiler: BaseQueryCompiler
```
Comment (YarShev) [inline modin/pandas/groupby.py:95]: Why do we need these as class attributes?

File: modin/pandas/iterator.py [hunk 6]
```diff
@@ -30,7 +36,9 @@ class PartitionIterator(Iterator):
         The function to get inner iterables from each partition.
     """
 
-    def __init__(self, df, axis, func):
+    df: DataFrame
```
Comment (YarShev) [inline modin/pandas/iterator.py:39]: Why do we need this as a class attribute?

File: modin/pandas/resample.py [hunk 7]
```diff
@@ -25,12 +27,19 @@
...
 
+if TYPE_CHECKING:
+    from modin.core.storage_formats import BaseQueryCompiler
+    from modin.pandas import DataFrame, Series
+
 
 @_inherit_docstrings(pandas.core.resample.Resampler)
 class Resampler(ClassLogger):
+    _dataframe: Union[DataFrame, Series]
+    _query_compiler: BaseQueryCompiler
```
Comment (YarShev) [inline modin/pandas/resample.py:38]: same

File: modin/pandas/series_utils.py [hunk 8]
```diff
@@ -30,10 +33,16 @@
...
 
+    from modin.core.storage_formats import BaseQueryCompiler
+    from modin.pandas import Series
+
 
 @_inherit_docstrings(pandas.core.arrays.arrow.ListAccessor)
 class ListAccessor(ClassLogger):
-    def __init__(self, data=None):
+    _series: Series
+    _query_compiler: BaseQueryCompiler
```
Comment (YarShev) [inline modin/pandas/series_utils.py:43]: same

File: modin/distributed/dataframe/pandas/partitions.py (hunk 1 above)
Comment (anmyachev) [reply modin/distributed/dataframe/pandas/partitions.py:93]: `mypy` requires that attribute `_modin_frame` be specified for `_query_compiler`.

File: modin/pandas/accessor.py (hunk 2 above)
Comment (anmyachev) [reply modin/pandas/accessor.py:218]: This is a practice that allows to specify exactly what attributes all instances have and, in addition, makes it possible to specify its type.

What is written here applies to all of your similar comments.

**Pull request merged**